File: docs/sdks/javascript/src/exceptions.ts
/**
 * API Lens JavaScript SDK - Custom Exceptions
 */

/**
 * Base exception for all API Lens errors
 */
export class APILensError extends Error {
  public readonly statusCode?: number;
  public readonly responseData?: Record<string, any>;
  public readonly requestId?: string;

  constructor(
    message: string,
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message);
    this.name = this.constructor.name;
    this.statusCode = statusCode;
    this.responseData = responseData || {};
    this.requestId = requestId;
    
    // Ensure the error is properly captured in stack traces
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }

  toString(): string {
    let message = this.message;
    if (this.statusCode) {
      message = `[${this.statusCode}] ${message}`;
    }
    if (this.requestId) {
      message = `${message} (Request ID: ${this.requestId})`;
    }
    return message;
  }
}

/**
 * Authentication failed - invalid or expired API key
 */
export class AuthenticationError extends APILensError {
  constructor(
    message: string = 'Authentication failed',
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
  }
}

/**
 * Authorization failed - insufficient permissions
 */
export class AuthorizationError extends APILensError {
  constructor(
    message: string = 'Insufficient permissions',
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
  }
}

/**
 * Request validation failed
 */
export class ValidationError extends APILensError {
  constructor(
    message: string = 'Request validation failed',
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
  }

  /**
   * Get validation error details from response
   */
  get validationErrors(): Record<string, any> {
    return this.responseData?.detail || {};
  }
}

/**
 * Resource not found
 */
export class NotFoundError extends APILensError {
  constructor(
    message: string = 'Resource not found',
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
  }
}

/**
 * Rate limit exceeded
 */
export class RateLimitError extends APILensError {
  public readonly retryAfter?: number;

  constructor(
    message: string = 'Rate limit exceeded',
    retryAfter?: number,
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
    this.retryAfter = retryAfter;
  }

  static fromResponse(
    responseData: Record<string, any>,
    statusCode?: number,
    requestId?: string
  ): RateLimitError {
    const retryAfter = responseData.retry_after;
    const limitType = responseData.limit_type || 'requests';
    const currentUsage = responseData.current_usage || 0;
    const limit = responseData.limit || 0;

    let message = `Rate limit exceeded for ${limitType}: ${currentUsage}/${limit}`;
    if (retryAfter) {
      message += `. Retry after ${retryAfter} seconds`;
    }

    return new RateLimitError(message, retryAfter, statusCode, responseData, requestId);
  }
}

/**
 * Usage quota exceeded
 */
export class QuotaExceededError extends APILensError {
  public readonly quotaType?: string;
  public readonly currentUsage?: number;
  public readonly quotaLimit?: number;

  constructor(
    message: string = 'Usage quota exceeded',
    quotaType?: string,
    currentUsage?: number,
    quotaLimit?: number,
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
    this.quotaType = quotaType;
    this.currentUsage = currentUsage;
    this.quotaLimit = quotaLimit;
  }

  static fromResponse(
    responseData: Record<string, any>,
    statusCode?: number,
    requestId?: string
  ): QuotaExceededError {
    const quotaType = responseData.quota_type || 'requests';
    const currentUsage = responseData.current_usage || 0;
    const quotaLimit = responseData.quota_limit || 0;

    const message = `Quota exceeded for ${quotaType}: ${currentUsage}/${quotaLimit}`;

    return new QuotaExceededError(
      message,
      quotaType,
      currentUsage,
      quotaLimit,
      statusCode,
      responseData,
      requestId
    );
  }
}

/**
 * Server-side error occurred
 */
export class ServerError extends APILensError {
  constructor(
    message: string = 'Internal server error',
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
  }
}

/**
 * API Lens service is temporarily unavailable
 */
export class ServiceUnavailableError extends ServerError {
  constructor(
    message: string = 'Service temporarily unavailable',
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
  }
}

/**
 * Error from AI vendor (OpenAI, Anthropic, etc.)
 */
export class VendorError extends APILensError {
  public readonly vendor?: string;
  public readonly vendorErrorCode?: string;
  public readonly vendorMessage?: string;

  constructor(
    message: string,
    vendor?: string,
    vendorErrorCode?: string,
    vendorMessage?: string,
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
    this.vendor = vendor;
    this.vendorErrorCode = vendorErrorCode;
    this.vendorMessage = vendorMessage;
  }

  static fromResponse(
    responseData: Record<string, any>,
    statusCode?: number,
    requestId?: string
  ): VendorError {
    const vendor = responseData.vendor;
    const vendorError = responseData.vendor_error || {};
    const vendorErrorCode = vendorError.code;
    const vendorMessage = vendorError.message;

    const message = `Vendor error from ${vendor}: ${vendorMessage || 'Unknown error'}`;

    return new VendorError(
      message,
      vendor,
      vendorErrorCode,
      vendorMessage,
      statusCode,
      responseData,
      requestId
    );
  }
}

/**
 * Error related to vendor API keys (BYOK)
 */
export class VendorKeyError extends APILensError {
  public readonly vendor?: string;
  public readonly keyIssue?: string;

  constructor(
    message: string,
    vendor?: string,
    keyIssue?: string,
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
    this.vendor = vendor;
    this.keyIssue = keyIssue;
  }
}

/**
 * Network connectivity error
 */
export class NetworkError extends APILensError {
  constructor(
    message: string = 'Network connection failed',
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
  }
}

/**
 * Request timeout
 */
export class TimeoutError extends NetworkError {
  public readonly timeout?: number;

  constructor(
    message: string = 'Request timed out',
    timeout?: number,
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
    this.timeout = timeout;
  }
}

/**
 * SDK configuration error
 */
export class ConfigurationError extends APILensError {
  constructor(
    message: string = 'SDK configuration error',
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
  }
}

/**
 * Data export operation failed
 */
export class ExportError extends APILensError {
  public readonly exportType?: string;
  public readonly exportFormat?: string;

  constructor(
    message: string = 'Export operation failed',
    exportType?: string,
    exportFormat?: string,
    statusCode?: number,
    responseData?: Record<string, any>,
    requestId?: string
  ) {
    super(message, statusCode, responseData, requestId);
    this.exportType = exportType;
    this.exportFormat = exportFormat;
  }
}

/**
 * Exception mapping for HTTP status codes
 */
export const STATUS_CODE_TO_EXCEPTION: Record<number, typeof APILensError> = {
  400: ValidationError,
  401: AuthenticationError,
  403: AuthorizationError,
  404: NotFoundError,
  422: ValidationError,
  429: RateLimitError,
  500: ServerError,
  502: ServiceUnavailableError,
  503: ServiceUnavailableError,
  504: TimeoutError,
};

/**
 * Create appropriate exception from HTTP response
 */
export function createExceptionFromResponse(
  statusCode: number,
  responseData: Record<string, any>,
  defaultMessage: string = 'API request failed'
): APILensError {
  // Get error details from response
  const errorMessage = responseData.detail || responseData.message || defaultMessage;
  const errorCode = responseData.error_code;
  const requestId = responseData.request_id;

  // Handle specific error types
  if (statusCode === 429) {
    return RateLimitError.fromResponse(responseData, statusCode, requestId);
  }

  if (responseData.vendor_error) {
    return VendorError.fromResponse(responseData, statusCode, requestId);
  }

  if (responseData.quota_exceeded) {
    return QuotaExceededError.fromResponse(responseData, statusCode, requestId);
  }

  // Use status code mapping
  const ExceptionClass = STATUS_CODE_TO_EXCEPTION[statusCode] || APILensError;

  return new ExceptionClass(errorMessage, statusCode, responseData, requestId);
}

/**
 * Check if an error is retryable
 */
export function isRetryableError(error: Error): boolean {
  if (error instanceof NetworkError || error instanceof TimeoutError || error instanceof ServerError) {
    return true;
  }

  if (error instanceof RateLimitError) {
    return true;
  }

  if (error instanceof APILensError && error.statusCode) {
    // Retry on server errors and rate limits
    return error.statusCode >= 500 || error.statusCode === 429;
  }

  return false;
}

/**
 * Calculate retry delay for an exception
 */
export function getRetryDelay(error: Error, attempt: number, baseDelay: number = 1000): number {
  if (error instanceof RateLimitError && error.retryAfter) {
    return error.retryAfter * 1000; // Convert to milliseconds
  }

  // Exponential backoff with jitter
  const delay = baseDelay * Math.pow(2, attempt);
  const jitter = Math.random() * 0.3 * delay;
  return delay + jitter;
}

File: docs/sdks/python/apilens/exceptions.py
"""
API Lens Python SDK - Custom Exceptions

This module defines custom exception classes for API Lens SDK operations.
"""

from typing import Optional, Dict, Any


class APILensError(Exception):
    """Base exception for all API Lens errors"""
    
    def __init__(
        self,
        message: str,
        status_code: Optional[int] = None,
        response_data: Optional[Dict[str, Any]] = None,
        request_id: Optional[str] = None
    ):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.response_data = response_data or {}
        self.request_id = request_id
    
    def __str__(self) -> str:
        base_msg = self.message
        if self.status_code:
            base_msg = f"[{self.status_code}] {base_msg}"
        if self.request_id:
            base_msg = f"{base_msg} (Request ID: {self.request_id})"
        return base_msg
    
    def __repr__(self) -> str:
        return f"{self.__class__.__name__}(message='{self.message}', status_code={self.status_code})"


class AuthenticationError(APILensError):
    """Authentication failed - invalid or expired API key"""
    
    def __init__(self, message: str = "Authentication failed", **kwargs):
        super().__init__(message, **kwargs)


class AuthorizationError(APILensError):
    """Authorization failed - insufficient permissions"""
    
    def __init__(self, message: str = "Insufficient permissions", **kwargs):
        super().__init__(message, **kwargs)


class ValidationError(APILensError):
    """Request validation failed"""
    
    def __init__(self, message: str = "Request validation failed", **kwargs):
        super().__init__(message, **kwargs)
        
    @property
    def validation_errors(self) -> Dict[str, Any]:
        """Get validation error details from response"""
        return self.response_data.get("detail", {})


class NotFoundError(APILensError):
    """Resource not found"""
    
    def __init__(self, message: str = "Resource not found", **kwargs):
        super().__init__(message, **kwargs)


class RateLimitError(APILensError):
    """Rate limit exceeded"""
    
    def __init__(
        self,
        message: str = "Rate limit exceeded",
        retry_after: Optional[int] = None,
        **kwargs
    ):
        super().__init__(message, **kwargs)
        self.retry_after = retry_after
    
    @classmethod
    def from_response(cls, response_data: Dict[str, Any], **kwargs):
        """Create RateLimitError from API response"""
        retry_after = response_data.get("retry_after")
        limit_type = response_data.get("limit_type", "requests")
        current_usage = response_data.get("current_usage", 0)
        limit = response_data.get("limit", 0)
        
        message = f"Rate limit exceeded for {limit_type}: {current_usage}/{limit}"
        if retry_after:
            message += f". Retry after {retry_after} seconds"
            
        return cls(message=message, retry_after=retry_after, **kwargs)


class QuotaExceededError(APILensError):
    """Usage quota exceeded"""
    
    def __init__(
        self,
        message: str = "Usage quota exceeded",
        quota_type: Optional[str] = None,
        current_usage: Optional[int] = None,
        quota_limit: Optional[int] = None,
        **kwargs
    ):
        super().__init__(message, **kwargs)
        self.quota_type = quota_type
        self.current_usage = current_usage
        self.quota_limit = quota_limit
    
    @classmethod
    def from_response(cls, response_data: Dict[str, Any], **kwargs):
        """Create QuotaExceededError from API response"""
        quota_type = response_data.get("quota_type", "requests")
        current_usage = response_data.get("current_usage", 0)
        quota_limit = response_data.get("quota_limit", 0)
        
        message = f"Quota exceeded for {quota_type}: {current_usage}/{quota_limit}"
        
        return cls(
            message=message,
            quota_type=quota_type,
            current_usage=current_usage,
            quota_limit=quota_limit,
            **kwargs
        )


class ServerError(APILensError):
    """Server-side error occurred"""
    
    def __init__(self, message: str = "Internal server error", **kwargs):
        super().__init__(message, **kwargs)


class ServiceUnavailableError(ServerError):
    """API Lens service is temporarily unavailable"""
    
    def __init__(self, message: str = "Service temporarily unavailable", **kwargs):
        super().__init__(message, **kwargs)


class VendorError(APILensError):
    """Error from AI vendor (OpenAI, Anthropic, etc.)"""
    
    def __init__(
        self,
        message: str,
        vendor: Optional[str] = None,
        vendor_error_code: Optional[str] = None,
        vendor_message: Optional[str] = None,
        **kwargs
    ):
        super().__init__(message, **kwargs)
        self.vendor = vendor
        self.vendor_error_code = vendor_error_code
        self.vendor_message = vendor_message
    
    @classmethod
    def from_response(cls, response_data: Dict[str, Any], **kwargs):
        """Create VendorError from API response"""
        vendor = response_data.get("vendor")
        vendor_error = response_data.get("vendor_error", {})
        vendor_error_code = vendor_error.get("code")
        vendor_message = vendor_error.get("message")
        
        message = f"Vendor error from {vendor}: {vendor_message or 'Unknown error'}"
        
        return cls(
            message=message,
            vendor=vendor,
            vendor_error_code=vendor_error_code,
            vendor_message=vendor_message,
            **kwargs
        )


class VendorKeyError(APILensError):
    """Error related to vendor API keys (BYOK)"""
    
    def __init__(
        self,
        message: str,
        vendor: Optional[str] = None,
        key_issue: Optional[str] = None,
        **kwargs
    ):
        super().__init__(message, **kwargs)
        self.vendor = vendor
        self.key_issue = key_issue


class NetworkError(APILensError):
    """Network connectivity error"""
    
    def __init__(self, message: str = "Network connection failed", **kwargs):
        super().__init__(message, **kwargs)


class TimeoutError(NetworkError):
    """Request timeout"""
    
    def __init__(self, message: str = "Request timed out", timeout: Optional[float] = None, **kwargs):
        super().__init__(message, **kwargs)
        self.timeout = timeout


class ConfigurationError(APILensError):
    """SDK configuration error"""
    
    def __init__(self, message: str = "SDK configuration error", **kwargs):
        super().__init__(message, **kwargs)


class ExportError(APILensError):
    """Data export operation failed"""
    
    def __init__(
        self,
        message: str = "Export operation failed",
        export_type: Optional[str] = None,
        export_format: Optional[str] = None,
        **kwargs
    ):
        super().__init__(message, **kwargs)
        self.export_type = export_type
        self.export_format = export_format


# Exception mapping for HTTP status codes
STATUS_CODE_TO_EXCEPTION = {
    400: ValidationError,
    401: AuthenticationError,
    403: AuthorizationError,
    404: NotFoundError,
    422: ValidationError,
    429: RateLimitError,
    500: ServerError,
    502: ServiceUnavailableError,
    503: ServiceUnavailableError,
    504: TimeoutError,
}


def create_exception_from_response(
    status_code: int,
    response_data: Dict[str, Any],
    default_message: str = "API request failed"
) -> APILensError:
    """Create appropriate exception from HTTP response"""
    
    # Get error details from response
    error_message = response_data.get("detail", response_data.get("message", default_message))
    error_code = response_data.get("error_code")
    request_id = response_data.get("request_id")
    
    # Handle specific error types
    if status_code == 429:
        return RateLimitError.from_response(
            response_data,
            status_code=status_code,
            response_data=response_data,
            request_id=request_id
        )
    
    if response_data.get("vendor_error"):
        return VendorError.from_response(
            response_data,
            status_code=status_code,
            response_data=response_data,
            request_id=request_id
        )
    
    if response_data.get("quota_exceeded"):
        return QuotaExceededError.from_response(
            response_data,
            status_code=status_code,
            response_data=response_data,
            request_id=request_id
        )
    
    # Use status code mapping
    exception_class = STATUS_CODE_TO_EXCEPTION.get(status_code, APILensError)
    
    return exception_class(
        message=error_message,
        status_code=status_code,
        response_data=response_data,
        request_id=request_id
    )


# Utility functions for exception handling
def is_retryable_error(exception: Exception) -> bool:
    """Check if an error is retryable"""
    if isinstance(exception, (NetworkError, TimeoutError, ServerError)):
        return True
    
    if isinstance(exception, RateLimitError):
        return True
    
    if isinstance(exception, APILensError) and exception.status_code:
        # Retry on server errors and rate limits
        return exception.status_code >= 500 or exception.status_code == 429
    
    return False


def get_retry_delay(exception: Exception, attempt: int, base_delay: float = 1.0) -> float:
    """Calculate retry delay for an exception"""
    if isinstance(exception, RateLimitError) and exception.retry_after:
        return float(exception.retry_after)
    
    # Exponential backoff with jitter
    import random
    delay = base_delay * (2 ** attempt)
    jitter = random.uniform(0.1, 0.3) * delay
    return delay + jitter